### DISCUSSION TITLE
Use Symbol in ES6 instead of String

### DISCUSSION THREAD

**Discussion by nwtgck** [opal/opal#2212] [General]
Hi, Opal team!

I'm new to Opal. This issue is just a proposal.

ES6 has `Symbol` so I think it may be better to use `Symbol` in ES6 instead of `String`.
MDN Document: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol

Here is an example code for Opal to use `Symbol` in ES6.
```rb
# Delete original `Symbol` which equals to `String`
Object.send(:remove_const, :Symbol)

# Define new Symbol using Symbol in JavaScript
class Symbol < `Symbol`
  def to_s
    ":"+`Symbol.keyFor(self.valueOf())`
  end
end

class String
  # Overwrite to_sym
  def to_sym
    `Symbol.for(self)`
  end
end


# Craete symbol (because :mysymbol1234 creates a string)
a = "mysymbol1234".to_sym
# Print symbol
p(a)
# => :mysymbol1234
```

You can try on Try Opal:
http://opalrb.com/try/?code:%23%20Delete%20original%20%60Symbol%60%20which%20equals%20to%20%60String%60%0AObject.send(%3Aremove_const%2C%20%3ASymbol)%0A%0A%23%20Define%20new%20Symbol%20using%20Symbol%20in%20JavaScript%0Aclass%20Symbol%20%3C%20%60Symbol%60%0A%20%20def%20to_s%0A%20%20%20%20%22%3A%22%2B%60Symbol.keyFor(self.valueOf())%60%0A%20%20end%0Aend%0A%0Aclass%20String%0A%20%20%23%20Overwrite%20to_sym%0A%20%20def%20to_sym%0A%20%20%20%20%60Symbol.for(self)%60%0A%20%20end%0Aend%0A%0A%0A%23%20Craete%20symbol%20(because%20%3Amysymbol1234%20creates%20a%20string)%0A%0Aa%20%3D%20%22mysymbol1234%22.to_sym%0A%23%20Print%20symbol%0Ap(a)%0A%23%20%3D%3E%20%3Amysymbol1234%0A#

Thank you!

**Comment by elia** [comment]
That would finally make symbols & strings two different things in opal (although it's quite useful at times! 😄). I did some quick research and perf test, and there are some issues to resolve before proceeding in this direction. Here's an IPS test:

```
=== Files: benchmark-ips/bm_js_symbols_vs_strings.rb
bundle exec opal --dynamic-require ignore --missing-require ignore -ropal/platform -gbenchmark-ips -rbenchmark/ips -A benchmark-ips/bm_js_symbols_vs_strings.rb
"/Users/elia/.rvm/gems/ruby-2.5.1/bin/opal"
Warming up --------------------------------------
       global symbol   144.765k i/100ms
              symbol   160.904k i/100ms
               ident   164.064k i/100ms
              string   164.388k i/100ms
Calculating -------------------------------------
       global symbol      4.863M (± 5.2%) i/s -     24.321M in   5.016262s
              symbol      5.297M (± 3.9%) i/s -     26.549M in   5.021029s
               ident      5.572M (± 4.7%) i/s -     27.891M in   5.017952s
              string      5.549M (± 5.8%) i/s -     27.782M in   5.026439s

Comparison:
               ident:  5571619.6 i/s
              string:  5548841.3 i/s - same-ish: difference falls within error
              symbol:  5297424.0 i/s - same-ish: difference falls within error
       global symbol:  4863483.6 i/s - 1.15x  slower

```

and the code: 

```rb
Benchmark.ips do |x|
  %x{
    var o = {}
    o[Symbol.for('foo')] = 123
    o.foo = 123
    var foo = Symbol('foo')
    o[foo] = 123
    var a = 0, b = 0, c = 0
  }
  
  x.report('global symbol') do
    `a += o[Symbol.for('foo')]`
  end

  x.report('symbol') do
    `a += o[foo]`
  end

  x.report('ident') do
    `b += o.foo`
  end

  x.report('string') do
    `c += o['foo']`
  end

  x.compare!
end
```

Looks like:
- `Symbol.for` is slower, might not be relevant, depending on how symbols would be used, but we should clarify that first
- not using `.for` would mean using an opal-only global registry

Also in the _cons_ symbols are much harder to use from the JS console

---

Also I found [this article](https://hacks.mozilla.org/2015/06/es6-in-depth-symbols/) very interesting, it explains that js symbols != ruby symbols, although we can still _abuse_ them for this purpose 👍 

_Feel free to start experimenting and please open a PR if you do, even if it's WIP / incomplete / broken_

**Comment by elia** [comment]
Also there's the problem of engine support that is still lacking in some areas http://kangax.github.io/compat-table/es6/#test-Symbol

**Comment by nwtgck** [comment]
@elia Thank you very much for the research and its result. The result tells me how slow symbols are because I don't know the performance difference between string and symbol in JavaScript. Thank you!

Since I don't know the culture, thought or history of Opal well, my proposal may not fit Opal. I just hope Opal was more close to real Ruby since Ruby's Symbol != Ruby's String. Could you give me what important things are and their priorities? (e.g. performance including speed or memory, Ruby compatibility or other things)

**Comment by elia** [comment]
@nwtgck no hard rules, especially if the proposal brings opal more close to ruby, it's often a compromise between performance, compliance, implementation etc. 

In some cases we sacrificed some performance for a greater gain in compliance, at other times the opposite was true. 

In this case my opinion is that 15% slower in a microbenchmark might not be that bad, especially if we come up with some trick to reduce or eliminate the slowdown in real-world code (like caching the symbols in local vars for each file or something like that.

**If you'd like to go forward with this I'll be happy to give you all the technical and historical support needed, would be great to have String != Symbol in Opal. 👍**
